packages/services/src/kernel/default.ts

```diff
@@ -851,11 +850,13 @@ export class DefaultKernel implements Kernel.IKernel {
   /**
    * Clear the internal state.
    */
-  private _clearState(): void {
+  private async _clearState(): Promise<void> {
     this._isReady = false;
     this._pendingMessages = [];
+    const futuresResolved: Promise<KernelMessage.IShellMessage>[] = [];
     this._futures.forEach(future => {
       future.dispose();
+      futuresResolved.push(future.done);
```

Comment:
How is this helping? The dispose should be synchronous. How is collecting these promises and then waiting on them collectively helping here?

Reply:
Many of the test failures we have seen (possibly most) have been due to hitting this thrown promise after a kernel shutdown/restart: https://github.com/jupyterlab/jupyterlab/blob/38c68660401dfd20cc3994a6af6e5f20f05fc24e/packages/services/src/kernel/future.ts#L176-L179

This was an attempt to more strictly make sure that those messages didn't leak. That being said, you have a much deeper knowledge of this part of the codebase than I, so I can try to back that out if it seems wrong.

Reply:
I saw this a lot as well. Do you happen to have a link where that failure happened?

Reply:
(though still, that catch *should* prevent the error from being thrown as an unhandled promise.)

Reply:
The test results of current `master` has several of these mode of failure: https://dev.azure.com/jupyterlab/jupyterlab/_build/results?buildId=636